transmart-core-db/grails-app/services/org/transmartproject/db/support/ParallelPatientSetTaskService.groovy

```diff
@@ -0,0 +1,136 @@
+import org.transmartproject.db.multidimquery.query.TrueConstraint
+
+import java.util.function.Function
+
+import static groovyx.gpars.GParsPool.withPool
+
+class ParallelPatientSetTaskService {
+
+    @Autowired
+    MultiDimensionalDataResource multiDimensionalDataResource
+
+    @Autowired
+    SystemResource systemResource
+
+    @Canonical
+    @CompileStatic
+    static class TaskParameters {
+        Constraint constraint
+        User user
+    }
+
+    @Canonical
+    @CompileStatic
+    static class SubtaskParameters {
+        int task
+        Constraint constraint
+        User user
+    }
+
+    @Canonical
+    @CompileStatic
+    static class ConstraintParts {
+        PatientSetConstraint patientSetConstraint
+        Constraint otherConstraint
+    }
+
+    @CompileStatic
+    static ConstraintParts getConstraintParts(Constraint constraint) {
+        PatientSetConstraint patientSetConstraint = null
+        Constraint otherConstraint = new TrueConstraint()
+        if (constraint instanceof PatientSetConstraint) {
+            patientSetConstraint = (PatientSetConstraint)constraint
+        } else if (constraint instanceof Combination && ((Combination)constraint).operator == Operator.AND) {
+            def groups = ((Combination)constraint).args.split { it instanceof PatientSetConstraint }
+            def patientSetConstraints = groups[0] as List<PatientSetConstraint>
+            def otherConjuncts = groups[1] as List<Constraint>
+            if (patientSetConstraints.size() == 1) {
+                patientSetConstraint = patientSetConstraints[0]
+                otherConstraint = new AndConstraint(otherConjuncts).normalise()
+            }
+        }
+        new ConstraintParts(patientSetConstraint, otherConstraint)
+    }
+
+    /**
+     * Splits a constraint into a patient set subconstraint and other subconstraint,
+     * and splits up the task into multiple parallel tasks that are applied to subsets of
+     * the patient set. When all are completed, the combine task is called.
+     * For parallelisation, the constraint needs to be a conjunction of a {@link PatientSetConstraint} and other constraints,
+     * or a singleton {@link PatientSetConstraint}.
+     * If the constraint cannot be split in this way, the sequentialTask fallback is used instead.
+     *
+     * @param constraint the constraint on which to apply the task.
+     * @param user
+     * @param sequentialTask the sequential fallback implementation.
+     * @param parallelSubTask the subtask implementation.
+     * @param combine the function that combines the results of the subtasks.
+     * @return the result of the combination of subtasks, or of the sequential fallback.
+     */
+    public <ResultType, SubtaskResultEntry> ResultType run(
+            TaskParameters parameters,
+            Function<TaskParameters, ResultType> sequentialTask,
+            Function<SubtaskParameters, List<SubtaskResultEntry>> parallelSubTask,
```

Comment:
Wouldn't `parallelSubTask` suffice?  `parallelSubTask` would be basically `sequentialTask` when number of tasks = `1`.

Reply:
Agreed. I will change it.